### PR TITLE
[GSoC'21] Morphing should change the object

### LINKED ISSUE #286: Morphing should change the object
**Is your feature request related to a problem? Please explain.**
Currently it's a bit hard to morph an object and then change the color of the morphed object afterwards. This is due to the fact that morphing doesn't really change the object. 

**Describe the solution you'd like**
I think we should consider actually changing the object function by the `morph_to` function. The problem here is that it would make it impossible to work with `liveview`. Therefore, we might want to have it inside the `ObjectSetting` instead such that we could overwrite the function without loosing the user defined function.

**Describe alternatives you've considered**
Another possibility would be that we can somehow return an object from `act!` but that feels weird as well.

### PR DESCRIPTION
**Pull request by Sov-trotter**
## PR Checklist

If you are contributing to `Javis.jl`, please make sure you are able to check off each item on this list:

- [x] Did I update `CHANGELOG.md` with whatever changes/features I added with this PR?
- [x] Did I make sure to only change the part of the file where I introduced a new change/feature?
- [x] Did I cover all corner cases to be close to 100% test coverage (if applicable)?
- [x] Did I properly add Javis dependencies to the `Project.toml` + set an upper bound of the dependency (if applicable)?
- [x] Did I properly add test dependencies to the `test` directory (if applicable)?
- [x] Did I check relevant tutorials that may be affected by changes in this PR?
- [x] Did I clearly articulate why this PR was made the way it was and how it was made?

**Link to relevant issue(s)**
Closes #286 

**How did you address these issues with this PR? What methods did you use?**
So the way this currently works is, by replacing the object's function with the function to which the object is supposed to be morphed.

So something like:
```julia
# inside the morph_to function
if frame == last(get_frames(action))
        object.func = to_func
    else
        # the morphing computation
        newpath()
        ....
        return morph_between(video, action, frame, from_polys, to_polys; do_action = do_action)
    end

```

[TEST SCRIPT](https://gist.github.com/Sov-trotter/6141ed9bed79bdb4a7b1a40b00227175)

### CODE REVIEW

File: test/viewer.jl [hunk 1]
```diff
@@ -95,7 +95,16 @@ end
     frames = Javis.preprocess_frames!(objects)
 
     @test v.filename === "foo.png"
-    img = Javis._pluto_viewer(vid, length(frames), objects)
+    img = Javis._pluto_viewer(vid, length(frames)-1, objects)
```
Comment (Wikunia) [inline test/viewer.jl:98]: does this mean that the pluto viewer will generally have problems? Like switching from the end frame of morphing back?
Maybe we should save the original function in the `Object` struct as well then and set it if it's not the last frame?

**Review by Wikunia — COMMENTED** [review]
Do you mind running `format` over this PR and make a note in the changelog?

File: test/viewer.jl (hunk 1 above)
Comment (Sov-trotter) [reply test/viewer.jl:98]: Yeah. I breaks jupyter and javis viewer. Pluto will work since we iterate over a list of frames.

File: test/viewer.jl [hunk 2]
```diff
@@ -95,7 +95,16 @@ end
     frames = Javis.preprocess_frames!(objects)
 
     @test v.filename === "foo.png"
-    img = Javis._pluto_viewer(vid, length(frames), objects)
+    img = Javis._pluto_viewer(vid, length(frames) - 1, objects)
```
Comment (Wikunia) [inline test/viewer.jl:98]: The old version should work now, right?

**Comment by Wikunia** [comment]
If it now works in Jupyter and the liveviewer I'm very happy with the PR. Then a single line in the `changelog` file would be nice I'll merge.

File: test/viewer.jl (hunk 2 above)
Comment (Sov-trotter) [reply test/viewer.jl:98]: Oh sorry. Forgot about that.

File: test/viewer.jl [hunk 3]
```diff
@@ -96,6 +96,9 @@ end
 
     @test v.filename === "foo.png"
     img = Javis._pluto_viewer(vid, length(frames), objects)
+    objects = vid.objects
+    frames = Javis.preprocess_frames!(objects)
```
Comment (Wikunia) [inline test/viewer.jl:100]: is this needed due to this PR?

Comment (Sov-trotter) [reply test/viewer.jl:100]: No. This is to test if the pluto viewer generates frames correctly.

Comment (Wikunia) [reply test/viewer.jl:100]: I was wondering why it wasn't there before

Comment (Sov-trotter) [reply test/viewer.jl:100]: I removed the black lines. https://github.com/Wikunia/Javis.jl/blob/master/test/viewer.jl#L94-L95

**Comment by Wikunia** [comment]
Great! Two things:
- `Changelog.md` needs the update
- If you would add a test case where this functionality is actually tested that would be awesome

**Comment by Sov-trotter** [comment]
So https://github.com/Wikunia/Javis.jl/pull/373/commits/6b5f4cf037853056b8efc803103de209c8588e37#diff-29df3566a6e16bde78716afd075fb04532e77f70ffad38d58cad1661b797f124R62 was creating a bug.

Now when we try to morph the object further, it restores the original function hence, if we morph a rect to a circ, then a circ to a star, it restores the rect when morphing for star has to begin since all of these actions act on the same object.

testmorph

**Comment by Sov-trotter** [comment]
The latest commit fixes this 
testmorph

File: src/morphs.jl [hunk 4]
```diff
@@ -42,6 +42,7 @@ function morph_to(to_func::Function; do_action = :stroke)
         _morph_to(video, object, action, frame, to_func; do_action = do_action)
 end
 
+const PREVIOUS_MORPH_ACTION = Array{Action,1}()
```
Comment (Wikunia) [inline src/morphs.jl:45]: I assume this isn't working when you more several objects at the same time.
I think this should be part of `opts` so per object

File: test/morphing.jl [hunk 5]
```diff
@@ -156,3 +156,63 @@ end
...
+    back = Background(1:200, ground)
+    star_obj = Object(1:200, abox)
+    act!(star_obj, Action(10:20, morph_to(acirc)))
+    act!(star_obj, Action(30:40, anim_translate(Point(100, -100))))
+
+    # this is also a bug with the morph_to function
+    # the star is formed at the incorrect position
+    # the the origin is shifted and neve restored back
+    # you can see this in the result of anim_translate actions
+    # fix this at some point!
```
Comment (Wikunia) [inline test/morphing.jl:180]: Maybe make this into a `TODO:` such that we find this at a later stage.

**Comment by Wikunia** [comment]
Is this ready from your side @Sov-trotter ?

**Comment by Sov-trotter** [comment]
Yeah!
